lib/types/array.js

```diff
@@ -661,28 +662,46 @@ internals.fillMissedErrors = function (schema, errors, requireds, value, state,
 };
 
 
-internals.checkRemainingOrdereds = function (schema, errors, ordereds, value, state, prefs) {
+internals.fillOrderedErrors = function (schema, errors, ordereds, value, state, prefs) {
 
     const requiredOrdereds = [];
-    const orderedsWithDefault = [];
 
     for (const ordered of ordereds) {
         if (ordered._flags.presence === 'required') {
             requiredOrdereds.push(ordered);
         }
-
-        const res = ordered.$_validate(undefined,state,prefs);
-        if (res.value !== undefined) {
-            orderedsWithDefault.push(res.value);
-        }
     }
 
     if (requiredOrdereds.length) {
         internals.fillMissedErrors(schema, errors, requiredOrdereds, value, state, prefs);
     }
+};
+
+
+internals.fillDefaults = function (ordereds, value, state, prefs) {
+
+    const defaultValues = [];
+    let trailingUndefined = true;
+
+    for (let i = ordereds.length - 1; i >= 0; --i) {
+
+        const ordered = ordereds[i];
+        const res = ordered.$_validate(undefined, state.localize(state.path, state.ancestors, ordered), prefs);
```

Comment:
You need to modify the path and ancestors, don't just relay them. That's why the references don't really resolve correctly (check my review below). Something like this would do 
```js
const key = wasArray ? i : new Number(i);       // eslint-disable-line no-new-wrappers (This bit is repeated somewhere else so extract to a function maybe?
const path = [...state.path, key];
const ancestors = [value, ...state.ancestors];

const res = ordered.$_validate(undefined, state.localize(path, ancestors, ordered), prefs);
```

Reply:
What a mistake, will be fixed in the next commit.

Reply:
I will mark this as resolved for now, until checking if there're changes needed when `Joi.link()` being tested.